perfkitbenchmarker/benchmarks/cloudsuite_web_serving.py

```diff
@@ -0,0 +1,368 @@
+                             'export FABAN_HOME=%s  '
+                             % (FABAN_HOME, BASE_DIR, BASE_DIR, BASE_DIR,
+                                OLIO_BUILD, OLIO_BUILD, OLIO_BUILD,
+                                FABAN_HOME))
+  client.RemoteCommand('perl -pi -e '
+                       '"s/CLIENT_IP/%s/g"'
+                       ' %s/run.xml' % (CLIENT_IP, FABAN_HOME))
+  client.RemoteCommand('perl -pi -e '
+                       '"s/FRONTEND_IP/%s/g"'
+                       ' %s/run.xml' % (FRONTEND_IP, FABAN_HOME))
+  client.RemoteCommand('perl -pi -e '
+                       '"s/MYSQL_DIR/%s/g" '
+                       '%s/run.xml' % (re.escape(MYSQL_HOME), FABAN_HOME))
+  client.RemoteCommand('perl -pi -e '
+                       '"s/OUTPUT_DIR/%s/g" '
+                       '%s/run.xml' % (re.escape(OUTPUT_DIR), FABAN_HOME))
+  filestore = posixpath.join(frontend.GetScratchDir(), 'filestore')
+  client.RemoteCommand('perl -pi -e '
+                       '"s/FILESTORE_DIR/%s/g" '
+                       '%s/run.xml' % (re.escape(filestore), FABAN_HOME))
+  client.RemoteCommand('perl -pi -e '
+                       '"s/BACKEND_IP/%s/g" '
+                       '%s/run.xml' % (BACKEND_IP, FABAN_HOME))
+
+
+def GetInfo():
+  return BENCHMARK_INFO
+
+
+def CheckPrerequisites():
+  """Verifies that the required resources are present.
+  Raises:
+    perfkitbenchmarker.data.ResourceNotFound: On missing resource.
+  """
+  if FLAGS['num_vms'].present and FLAGS.num_vms < 3:
+    raise ValueError('Web Serving requires at least 3 VMs')
+
+
+def PreparePrivateKey(vm):
+  vm.AuthenticateVm()
+
+
+def PrepareVms(vm):
+  vm.Install('wget')
+  vm.Install('ant')
+  vm.Install('openjdk7')
+  vm.RemoteCommand('cd %s && '
+                   'wget parsa.epfl.ch/cloudsuite/software/web.tar.gz && '
+                   'tar xzf web.tar.gz' % vm_util.VM_TMP_DIR)
+
+
+def Prepare(benchmark_spec):
+  """Install Java, apache ant, authenticate vms
+     Set up the client machine, backend machine, and frontend
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+  """
+  vms = benchmark_spec.vms
+  vm_util.RunThreaded(PreparePrivateKey, vms)
+  client = vms[CL]
+  backend = vms[BK]
+  frontend = vms[FR]
+  vm_util.RunThreaded(PrepareVms, vms)
+  frontend.RemoteCommand('cd %s && '
+                         'wget parsa.epfl.ch/cloudsuite/software/perfkit/'
+                         'web_serving/webservingfiles.tgz && '
+                         'tar xzf webservingfiles.tgz' % BASE_DIR)
+  setupClient(benchmark_spec)
+  frontend.RemoteCommand('scp -r -o StrictHostKeyChecking=no %s:%s %s' %
+                         (client.ip_address, FABAN_HOME, BASE_DIR))
+  backend.RemoteCommand('scp -r -o StrictHostKeyChecking=no %s:%s %s' %
+                        (client.ip_address, FABAN_HOME, BASE_DIR))
+  setupBackend(benchmark_spec)
+  setupFrontend(benchmark_spec)
```

Comment:
Alternately, you do have those VM IDs defined at the top of the file (CL, BK, FR). You could have a function like `_SetupVM(benchmark_spec, thread_id)` that then selects which setup function to call based on the passed identifier. This might be cleaner than my previous suggestion, as it is a benchmark-specific function that makes sense in the current module.